File: scratch/sort.lua
---table(배열) 을 출력
---@param t table
local function printArr(t)
    local sz = #t
    io.write("{ ")
    if sz >= 1 then
        io.write(t[1])
    end
    for i=2, sz do
        io.write(string.format(", %d", t[i]))
    end
    print(" }")
end


---정수 랜덤배열 생성
---@param size integer
---@param max integer
---@return table
local function getRandArray(size, max)
    local result = {}
    for i=1, size do
        table.insert(result, math.random(max))
    end
    return result
end


--- start index, end index가 멀쩡한 상태인지 점검하는 함수
--- @param sz integer
--- @param si integer
--- @param ei integer
--- @return boolean
local function checkIdx(sz, si, ei)
    if si == ei or si >= ei or ei - si + 1 > sz then
        print("index error")
        return false
    end
    if ei - si + 1 == sz and si ~= 1 then
        print("out of idx")
        return false
    end
    return true
end


---comment
---@param arr table
---@param si integer
---@param ei integer
local function bubble(arr, si, ei)
    if not checkIdx(#arr, si, ei) then
        return;
    end

    for i=si, ei-1 do
        for j=si, ei-i do
            if arr[j] > arr[j + 1] then
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
            end
        end
    end
end


--메인실행코드 함수화
local function main()

    local sampleSize = 10;
    local sample = getRandArray(sampleSize, 50)
    printArr(sample)
    local startTime = os.clock()
    bubble(sample, 1, sampleSize)
    print(string.format("sorting time: %.6f\n", os.clock() - startTime))
    printArr(sample)

end

math.randomseed(os.time() + os.clock() * 1000000)
main()

File: scratch/sort.dart
import 'dart:math';
import 'dart:typed_data';

void main(List<String> args) {

  // sortBench(quick, 100_000_000, useInt32List: true);
  sortBench(quick, 10, useInt32List: true, printArray: true);

}


// 정렬 벤치마킹
void sortBench(sortfunc sf, int sampleSize, { required bool useInt32List, bool printArray = false}) {
  // sample의 자료형은 런타임환경에 결정된다.
  DateTime initArrayStart = DateTime.now();
  dynamic sample = useInt32List? getRandInt32Array(sampleSize, 10000) : getRandList(sampleSize, 10000);
  DateTime initArrayEnd = DateTime.now();

  if (printArray && sampleSize < 50)    // printArray: true라도, 50개가 넘는 길이는 출력하지 않도록 강제한다.
    print(sample);
  DateTime start = DateTime.now();
  sort(sample, sf);
  DateTime res = DateTime.now();
  if (printArray && sampleSize < 50)
    print(sample);
  print("sort: ${res.difference(start).inMicroseconds / 1_000_000} \tinit sample: ${initArrayEnd.difference(initArrayStart).inMicroseconds / 1_000_000}");
}



// 정렬을 선택 할 수 있게 alias.
typedef void sortfunc(var arr, int start, int end);
void sortRange(var arr, int start, int end, sortfunc sf) {
  sf(arr, start, end);
}
void sort(var arr, sortfunc sf) {
  sortRange(arr, 0, arr.length - 1, sf);
}


void bubble(var arr, int start, int end) {
  for (int i = start; i <= end - 1; ++i) {
    for (int j = start; j <= end - 1 - i; ++j) {
      if (arr[j] > arr[j + 1]) {
        int tmp = arr[j];
        arr[j] = arr[j + 1];
        arr[j + 1] = tmp;
      }
    }
  }
}


void selection(var arr, int start, int end) {
  for (int i = start; i <= end; ++i) {
    int bfIdx = i;
    for (int j = i; j <= end; ++j) {
      if (arr[j] < arr[bfIdx]) {
        bfIdx = j;
      }
    }
    if (bfIdx != i) {
      int tmp = arr[i];
      arr[i] = arr[bfIdx];
      arr[bfIdx] = tmp;
    }
  }
}


void insertion(var arr, int start, int end) {
  for (int i = start + 1; i <= end; ++i) {
    int bf = arr[i];
    int j = i;
    for (; j > start && arr[j - 1] > bf; --j) {
      arr[j] = arr[j - 1];
    }
    if (j != i) {
      arr[j] = bf;
    }
  }
}


void shell(var arr, int start, int end) {
  int gap = 1;
  while (gap < (arr.length / 3).toInt()) {
    gap = gap * 3 + 1;
  }

  while (gap > 0) {
    for (int i = start + gap; i <= end; ++i) {
      int bf = arr[i];
      int j = i;
      while (j >= gap && arr[j - gap] > bf) {
        arr[j] = arr[j - gap];
        j -= gap;
      }
      if (j != i) {
        arr[j] = bf;
      }
    }
    gap = (gap / 3).toInt();
  }

}


void quick(var arr, int start, int end) {
  if (start >= end) return;
  int sz = end - start + 1;
  if (sz <= 256) {
    insertion(arr, start, end);
    return;
  }

  int left = start;
  int right = end;
  int pivot = arr[(start + (sz / 2).toInt()) - 1];

  while (left <= right) {
    while (arr[left] < pivot) left++;
    while (arr[right] > pivot) right--;
    if (left <= right) {
      int tmp = arr[left];
      arr[left] = arr[right];
      arr[right] = tmp;
      left++;
      right--;
    }
  }

  if (right < end) quick(arr, start, right);
  if (left > start) quick(arr, left, end);

}


// Int32List 자료형: 전통적인 int* 혹은 int[] 등의 배열과 유사
Int32List getRandInt32Array(int size, int max) {
  Int32List res = Int32List(size);
  Random rand = Random();
  for (int i = 0; i < size; ++i)
    res[i] = rand.nextInt(max);
  return res;
}

// 가변크기 배열(std::vector와 유사)
List<int> getRandList(int size, int max) {
  // add 방식으로 사용할 때 배열생성시간(20~30%), 정렬(2~ 3%) 정도 느림

  // List<int> res = [];                          // add 방식으로 사용할 때
  List<int> res = List.filled(size, 0);         // index기반으로 사용할 때
  Random rand = Random();
  for (int i = 0; i < size; ++i)
    // res.add(rand.nextInt(max));
    res[i] = rand.nextInt(max);
  return res;
}